addons/pos_viva_wallet/static/src/app/payment_viva_wallet.js

```diff
@@ -0,0 +1,199 @@
+    set_most_recent_service_id(id) {
+        this.most_recent_service_id = id;
+    }
+    pending_viva_wallet_line() {
+        return this.pos.getPendingPaymentLine("viva_wallet");
+    }
+
+    _handle_odoo_connection_failure(data = {}) {
+        // handle timeout
+        var line = this.pending_viva_wallet_line();
+        if (line) {
+            line.set_payment_status("retry");
+        }
+        this._show_error(
+            _t(
+                "Could not connect to the Odoo server, please check your internet connection and try again."
+            )
+        );
+
+        return Promise.reject(data); // prevent subsequent onFullFilled's from being called
+    }
+
+    _call_viva_wallet(data) {
+        return this.env.services.orm.silent
+            .call("pos.payment.method",
+                "viva_wallet_send_payment_request",
+                [[this.payment_method.id], data]
+            )
+            .catch(this._handle_odoo_connection_failure.bind(this));
+    }
+
+    _viva_wallet_handle_response(response) {
+        var line = this.pending_viva_wallet_line();
+        line.set_payment_status("waitingCard");
+        return this.waitForPaymentConfirmation();
+    }
+
+    _viva_wallet_pay () {
+        /**
+         * Override
+         */
+        super.send_payment_request(...arguments);
+        var order = this.pos.get_order();
+        var line = order.selected_paymentline;
+        let customerTrns = ' ';
+        line.set_payment_status("waitingCard");
+
+        if (line.amount < 0) {
+            this._show_error(_t("Cannot process transactions with negative amount."));
+            return false;
+        }
+
+        if (order.partner) {
+            customerTrns = order.partner.name + ' - ' + order.partner.email
+        }
+
+        line.sessionId = order.uid + ' - ' + uuidv4();
+        var data = {
+            "sessionId": line.sessionId,
+            "terminalId": line.payment_method.viva_wallet_terminal_id,
+            "cashRegisterId": this.pos.get_cashier().name,
+            "amount": line.amount * 100,
+            "currencyCode": 978, // Viva wallet only uses EUR 978 need add a new field numeric_code in res.currency
+            "merchantReference": line.sessionId + '/' + this.pos.pos_session.id,
+            "customerTrns": customerTrns,
+            "preauth": false,
+            "maxInstalments": 0,
+            "tipAmount": 0
+        };
+        return this._call_viva_wallet(data).then((data) => {
+            return this._viva_wallet_handle_response(data);
+        });
+    }
+
+    _viva_wallet_cancel (order, cid) {
+        /**
+         * Override
+         */
+        super.send_payment_cancel(...arguments);
+        const line = this.pos.get_order().selected_paymentline;
+
+        var data = {
+            "sessionId": line.sessionId,
+            "cashRegisterId": this.pos.get_cashier().name
+        };
+        try {
+            const resp = this.pos.env.services.orm.silent.call(
```

Comment:
```suggestion
            const resp = await this.pos.env.services.orm.silent.call(
```

`resp` here is a `Promise`. It will never have an "error" field. I should be awaited. But, you can only await inside an `async function`, so you need to put `async` in the declaration of `_viva_wallet_cancel`.

Also, this method doesn't seem to be compliant with the return of `viva_wallet_send_payment_cancel`. `viva_wallet_send_payment_cancel` return "{ 'success': code }".